core/src/main/java/com/google/cloud/sql/core/LazyRefreshStrategy.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Instant;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** RefreshStrategy that implements the lazy refresh strategy. */
+public class LazyRefreshStrategy implements RefreshStrategy {
+  private final Logger logger = LoggerFactory.getLogger(LazyRefreshStrategy.class);
+
+  private final String name;
+  private final Supplier<ConnectionInfo> refreshOperation;
+
+  private final Object connectionInfoGuard = new Object();
+
+  @GuardedBy("connectionInfoGuard")
+  private ConnectionInfo connectionInfo;
+
+  @GuardedBy("connectionInfoGuard")
+  private boolean closed;
+
+  /** Creates a new LazyRefreshStrategy instance. */
+  public LazyRefreshStrategy(String name, Supplier<ConnectionInfo> refreshOperation) {
+    this.name = name;
+    this.refreshOperation = refreshOperation;
+  }
+
+  @Override
+  public ConnectionInfo getConnectionInfo(long timeoutMs) {
+    synchronized (connectionInfoGuard) {
+      if (closed) {
+        throw new IllegalStateException("Named connection closed");
+      }
+
+      if (connectionInfo == null) {
+        logger.debug(
+            String.format(
+                "[%s] Client certificate has expired. Starting next refresh operation immediately.",
+                name));
+        fetchConnectionInfo();
+      }
```

Comment:
when does this case occur?

Reply:
looks like it occurs when we invalidate the cache, can we have a better debug message?

Reply:
maybe "Cache is empty. Starting next refresh immediately."

Reply:
Updated